Update eslint config in packages/runtime/runtime-utils to extend "recommended" base config

RishhiB opened the pull request:
The package currently uses our "minimal" config. This PR updates it to use "recommended" and fix the resulting linter violations.

[AB#3022](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/3022)

RishhiB committed:
Update eslint config in packages/runtime/runtime-utils to extend "recommended" base config

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

RishhiB committed:
breaking types

RishhiB committed:
other package changes

RishhiB committed:
undo utf8 changes

RishhiB committed:
undo more utf-8 changes

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

alexvy86 reviewed (commented):
Still have a few files to go through but leaving some comments for now.

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/test/handles.spec.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/test/handles.spec.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

RishhiB commented on packages/runtime/container-runtime/src/channelCollection.ts:
```diff
@@ -406,7 +406,7 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
 			// We need to process the GC Data for both local and remote attach messages
 			const foundGCData = processAttachMessageGCData(
-				attachMessage.snapshot,
+				attachMessage.snapshot ?? undefined,
```
`processAttachMessageGCData` takes param `snapshot` of type `ITree` or `undefined` but `attachMessage.snapshot` is type `ITree` or `null`
```ts
function processAttachMessageGCData(snapshot: ITree | undefined, addedGCOutboundRoute: (fromNodeId: string, toPath: string) => void): boolean
```

```ts
export type InboundAttachMessage = Omit<IAttachMessage, "snapshot"> & {
	// eslint-disable-next-line @rushstack/no-new-null -- TODO: breaking change; protocol might even explicitly use null
	snapshot: IAttachMessage["snapshot"] | null;
};```

alexvy86 replied:
Oh, I hadn't gotten to the fact that this PR changes the function signature too 😄. The function changes seem fine since nothing inside the function has special casing for `snapshot === null`... so yeah, I think this is fine. Double checking, @agarwal-navin and @markfields, any concerns here?

RishhiB committed:
helper function

RishhiB committed:
format

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

alexvy86 commented on packages/runtime/runtime-utils/src/requestParser.ts:
```diff
@@ -17,21 +17,25 @@ export class RequestParser implements IRequest {
+		const pathParts: string[] = [];
+		const urlPath = url.slice(
+			0,
+			Math.max(0, queryStartIndex < 0 ? url.length : queryStartIndex),
```
I think the `Math.max()` here is unnecessary? I get it's so `slice()` doesn't get a negative argument in case `queryStartIndex === -1`, but in this case the ternary already handles that.
```suggestion
			queryStartIndex < 0 ? url.length : queryStartIndex,
```

alexvy86 commented on packages/runtime/runtime-utils/src/summaryUtils.ts:
```diff
@@ -455,15 +479,16 @@ export function processAttachMessageGCData(
 	}
 
 	assert(
+		// eslint-disable-next-line unicorn/text-encoding-identifier-case
```
This is the one you mentioned where DDS use "utf-8" explicitly, right? If so let's add that to the all the disables in this file, since this is not a lint rule we'll be actively trying to remove, it'll probably stay for a while and we'll want to know why.

alexvy86 replied:
Bump on this comment.

jason-ha replied:
Disables should be commented

alexvy86 commented on packages/runtime/runtime-utils/src/summaryUtils.ts:
```diff
@@ -59,26 +59,35 @@ export function mergeStats(...stats: ISummaryStats[]): ISummaryStats {
 /**
+ * Gets the size of a blob
+ * @param content - The content of the blob
+ * @returns The size of the blob
```
```suggestion
 * @returns The size of the blob in bytes
```

alexvy86 commented on packages/runtime/runtime-utils/src/summaryUtils.ts:
```diff
@@ -59,26 +59,35 @@ export function mergeStats(...stats: ISummaryStats[]): ISummaryStats {
-		if (code >= 0xdc00 && code <= 0xdfff) {
-			i--; // trail surrogate
+		const code = str.codePointAt(i);
+		// Add null check to handle potential undefined
```
The comment feels unnecessary (unless there's a special condition that would cause it to be undefined that we need to explain, for example), the code is pretty clear about it.
```suggestion
```

alexvy86 commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -72,20 +94,29 @@ export function responseToException(response: IResponse, request: IRequest): Err
+ * @param value - Error message or description
+ * @param request - The request that resulted in this error
+ * @param headers - Optional headers to include in the response
+ * @returns An IResponse object representing the error
```
```suggestion
 * @returns An {@link @fluidframework/core-interfaces#IResponse } object representing the error
```

alexvy86 commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -111,6 +142,11 @@ export function createResponseError(
 export type Factory = IFluidDataStoreFactory & Partial<IProvideFluidDataStoreRegistry>;
 
 /**
+ * Creates a combined IFluidDataStoreFactory and IFluidDataStoreRegistry implementation
```
```suggestion
 * Creates a combined {@link @fluidframework/runtime-definitions#IFluidDataStoreFactory } and {@link @fluidframework/runtime-definitions#IFluidDataStoreRegistry } implementation
```

alexvy86 commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -72,20 +94,29 @@ export function responseToException(response: IResponse, request: IRequest): Err
 /**
+ * Creates a 404 "not found" response for the given request
+ * @param request - The request that resulted in the 404 response
+ * @returns An IResponse object with 404 status code
```
```suggestion
 * @returns An {@link @fluidframework/core-interfaces#IResponse } object with 404 status code
```

alexvy86 commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -17,45 +17,67 @@ interface IResponseException extends Error {
 
 /**
+ * Converts an IResponse object back into an Error object that can be thrown
+ * @param response - The IResponse object to convert
```
```suggestion
 * @param response - The {@link @fluidframework/core-interfaces#IResponse } object to convert
```

alexvy86 commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -17,45 +17,67 @@ interface IResponseException extends Error {
 /**
+ * Converts an IResponse object back into an Error object that can be thrown
+ * @param response - The IResponse object to convert
+ * @param request - The original IRequest object
```
```suggestion
 * @param request - The original {@link @fluidframework/core-interfaces#IRequest } object
```

alexvy86 commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -17,45 +17,67 @@ interface IResponseException extends Error {
 }
 
 /**
+ * Converts an IResponse object back into an Error object that can be thrown
```
```suggestion
 * Converts an {@link @fluidframework/core-interfaces#IResponse } object back into an Error object that can be thrown
```

alexvy86 commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -17,45 +17,67 @@ interface IResponseException extends Error {
 }
 
 /**
+ * Type guard for determining if an error is an IResponseException
```
```suggestion
 * Type guard for determining if an error is an {@link IResponseException }
```

alexvy86 reviewed (commented):
Mostly docs and similar. I'd like to hear back from the summarization crew before approving :)

RishhiB committed:
Update packages/runtime/runtime-utils/src/requestParser.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/summaryUtils.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/summaryUtils.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
format

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

Josmithr commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -17,45 +17,67 @@ interface IResponseException extends Error {
 	const status = 500;
-	if (err !== null && typeof err === "object" && err.errorFromRequestFluidObject === true) {
-		const responseErr: IResponseException = err;
+	if (isResponseException(err)) {
```
Nit (while we're here): can we call this `error`? That would be required if we upgrade the config to "strict" anyways.

Josmithr commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -17,45 +17,67 @@ interface IResponseException extends Error {
 }
 
 /**
+ * Converts an {@link @fluidframework/core-interfaces#IResponse } object back into an Error object that can be thrown
```
```suggestion
 * Converts an {@link @fluidframework/core-interfaces#IResponse} back into an Error object that can be thrown.
```

Josmithr commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -17,45 +17,67 @@ interface IResponseException extends Error {
+}
+
+/**
+ * Converts an error object into an {@link @fluidframework/core-interfaces#IResponse}
```
```suggestion
 * Converts an error object into an {@link @fluidframework/core-interfaces#IResponse}.
```

Josmithr commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -17,45 +17,67 @@ interface IResponseException extends Error {
 /**
+ * Converts an {@link @fluidframework/core-interfaces#IResponse } object back into an Error object that can be thrown
+ * @param response - The {@link @fluidframework/core-interfaces#IResponse } object to convert
+ * @param request - The original {@link @fluidframework/core-interfaces#IRequest } object
```
```suggestion
 * @param request - The original {@link @fluidframework/core-interfaces#IRequest}.
```

Josmithr commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -72,20 +94,29 @@ export function responseToException(response: IResponse, request: IRequest): Err
 /**
+ * Creates a 404 "not found" response for the given request
+ * @param request - The request that resulted in the 404 response
+ * @returns An {@link @fluidframework/core-interfaces#IResponse } object with 404 status code
```
```suggestion
 * @returns An {@link @fluidframework/core-interfaces#IResponse} with 404 status code.
```

Josmithr commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -72,20 +94,29 @@ export function responseToException(response: IResponse, request: IRequest): Err
+ * @param value - Error message or description
+ * @param request - The request that resulted in this error
+ * @param headers - Optional headers to include in the response
+ * @returns An {@link @fluidframework/core-interfaces#IResponse } object representing the error
```
```suggestion
 * @returns An {@link @fluidframework/core-interfaces#IResponse} representing the error
```

Josmithr reviewed (commented):
Left a few suggestions, and it looks like you have some conflicts to resolve. But overall looking good! Thanks for tackling this!

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

RishhiB committed:
update to error

alexvy86 commented on packages/runtime/runtime-utils/src/handles.ts:
```diff
@@ -23,8 +23,11 @@ export interface ISerializedHandle {
-export const isSerializedHandle = (value: any): value is ISerializedHandle =>
-	value?.type === "__fluid_handle__";
+export const isSerializedHandle = (value: unknown): value is ISerializedHandle =>
+	typeof value === "object" &&
+	value !== null &&
+	"type" in value &&
+	value.type === "__fluid_handle__";
```
Bump on this one too.

jason-ha replied:
API is one thing and internal treatment is another.
The prior code looked safe to me. (Someone should confirm that if a primitive like a string were passed in then it works correctly (before change).) So, correction should be to internal typing. In these situations we can use an `as` cast. `as { type?: string }` would be nice, but `as { type?: string } | undefined | null` might be more appropriate.
Ideally though we would limit the argument type to something reasonable like `object | undefined`. Is that not possible? What cases does it need to handle?

alexvy86 replied:
> Someone should confirm that if a primitive like a string were passed in then it works correctly (before change)

A string wouldn't have a `type` property so it should correctly return false in that case?

At a glance, the function seems to be used to process op contents, and to deserialize data stores. Not clear to me if it'd be easy to add type safety (or how much) in those scenarios. I imagine it's acceptable for a string (or any primitive) to be passed to the function and return false, which typing the input as `object | undefined` would prevent, right? I'd say for now keep the `unknown` in the API but simplify the implementation to use `as { type?: string } | undefined | null` instead of the extra checks. Seems like an improvement over `any` and since it's `@internal` we can later improve it more if necessary.

alexvy86 reviewed (commented):
Two last things from me and I think it looks good after that.

RishhiB committed:
alejandro's feedback

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

RishhiB committed:
shorten comment

jason-ha commented on packages/runtime/runtime-utils/src/requestParser.ts:
```diff
@@ -41,7 +42,8 @@ export class RequestParser implements IRequest {
 	protected constructor(private readonly request: Readonly<IRequest>) {
 		const queryStartIndex = this.request.url.indexOf("?");
-		this.query = queryStartIndex >= 0 ? this.request.url.substring(queryStartIndex) : "";
+		this.query =
+			queryStartIndex >= 0 ? this.request.url.slice(Math.max(0, queryStartIndex)) : "";
```
What is the complaint? queryStartIndex is already >= 0; so, `max(0,` isn't doing anything.

jason-ha commented on packages/runtime/runtime-utils/src/runtimeFactoryHelper.ts:
```diff
@@ -15,7 +15,7 @@ import { IContainerRuntime } from "@fluidframework/container-runtime-definitions
 export abstract class RuntimeFactoryHelper<T = IContainerRuntime> implements IRuntimeFactory {
-	public get IRuntimeFactory() {
+	public get IRuntimeFactory(): this {
 		return this;
```
`this` or `IRuntimeFactory`?

RishhiB replied:
when changing to `IRuntimeFactory`, im getting backward compat errors

jason-ha commented on packages/runtime/runtime-utils/src/summaryUtils.ts:
```diff
@@ -354,7 +375,7 @@ export function convertSnapshotTreeToSummaryTree(
 			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
 			const content: ArrayBufferLike = snapshot.blobsContents[id]!;
 			if (content !== undefined) {
-				decoded = bufferToString(content, "utf-8");
+				decoded = bufferToString(content, "utf8");
```
Why is this an okay change, but there isn't a similar change on line 415?
```diff
+				// eslint-disable-next-line unicorn/text-encoding-identifier-case
				let encoding: "utf-8" | "base64" = "utf-8";
				if (typeof value.content === "string") {
```

RishhiB replied:
It might not be ok, reverting this. thank you for the catch

alexvy86 replied:
@RishhiB seems like latest changes might not be pushed yet? This still says "utf8", and the lint disables still don't have comments about why they are safe/necessary.

jason-ha commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -17,45 +17,67 @@ interface IResponseException extends Error {
 export function responseToException(response: IResponse, request: IRequest): Error {
-	const message = response.value;
+	const message = response.value as string;
 	const errWithStack = generateErrorWithStack();
```
We need a rule that `as` should be commented. No other option? Why is this okay?

alexvy86 replied:
For these PRs that update the eslint config we've been leaning towards being conservative in the changes we make, especially since we're usually not experts in the packages. If a package is currently doing something weird, if it's evident that it's a bug and the fix is obvious then sure, let's fix it, but otherwise I think I'd suggest adding a comment for someone more familiar with the code to eventually address it and move on. But I agree that a comment is necessary.

jason-ha commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -17,45 +17,67 @@ interface IResponseException extends Error {
-			headers: responseErr.underlyingResponseHeaders,
+			headers: error.underlyingResponseHeaders,
 		};
 	}
 
-	// Capture error objects, not stack itself, as stack retrieval is very expensive operation, so we delay it
+	// Capture error objects, not stack itself, as stack retrieval is very expensive operation
 	const errWithStack = generateErrorWithStack();
 
 	return {
 		mimeType: "text/plain",
 		status,
-		value: `${err}`,
+		value: `${error}`,
 		get stack() {
-			return (err?.stack as string | undefined) ?? errWithStack.stack;
+			// Use type assertion after checking if error is an object with stack
+			return (
+				(typeof error === "object" && error !== null && "stack" in error
+					? (error.stack as string | undefined)
+					: undefined) ?? errWithStack.stack
```
Same question as for [isSerializedHandle ](https://github.com/microsoft/FluidFramework/pull/23956/files#r1987863430)

RishhiB replied:
Added comment explaining it's safe since we're only checking for property existence, if theres a better on please let me know

jason-ha replied:
Look at how isSerializedHandle was resolved in the end. This should go back to simpler logic like that. We don't need funny examples sitting around nor the [slightly] bloated code.

jason-ha commented on packages/common/driver-definitions/src/protocol/storage.ts:
```diff
@@ -46,7 +46,7 @@ export interface IBlob {
 	 * The encoding of the contents string
 	 */
-	// eslint-disable-next-line unicorn/text-encoding-identifier-case
+	// eslint-disable-next-line unicorn/text-encoding-identifier-case -- Not removing this rule as it would require a breaking change in driver-definitions
```
Can we have a better reason? Just because something is breaking doesn't mean it can't or shouldn't happen.

RishhiB committed:
feedback

RishhiB committed:
change from this to IRuntimeFactory

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

RishhiB committed:
add comments

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

RishhiB committed:
weird nit for file i didnt edit?

RishhiB committed:
Fix build errors in runtime-utils package

RishhiB committed:
Merge remote-tracking branch 'origin/main' into UpdateESlintToRecommendedRuntimeUtils

RishhiB committed:
Add comments for type assertions per PR feedback

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

RishhiB committed:
policy check

alexvy86 commented on packages/runtime/runtime-utils/src/remoteFluidObjectHandle.ts:
```diff
@@ -54,6 +54,7 @@ export class RemoteFluidObjectHandle extends FluidHandleBase<FluidObject> {
 			};
 			this.objectP = this.routeContext.resolveHandle(request).then<FluidObject>((response) => {
 				if (response.mimeType === "fluid/object") {
+					// Type assertion is safe due to mimeType check ensuring value is a FluidObject
```
A bit clearer; checking the mime type on its own is not really a guarantee.
```suggestion
					// Responses with mimeType == "fluid/object" are produced by Fluid, and thus they can reasonably be expected to always be a FluidObject
```

alexvy86 commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -12,45 +12,68 @@ interface IResponseException extends Error {
  */
 export function responseToException(response: IResponse, request: IRequest): Error {
-	const message = response.value;
+	// IResponse.value is typed as 'any', but we expect it to be a string error message
```
I'm not sure we really _expect_ it to. I think we should rather be conservative with what we think is happening in code we don't own.
```suggestion
	// As of 2025-08-20 the code seems to assume `response.value` is always a string.
	// This type assertion just encodes that assumption as we move to stricter linting rules, but it might need to be revisited.
```

RishhiB committed:
Update packages/runtime/runtime-utils/src/remoteFluidObjectHandle.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Update packages/runtime/runtime-utils/src/dataStoreHelpers.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

alexvy86 reviewed (approved):
Looks good as far as I can tell.

RishhiB committed:
Merge branch 'main' into UpdateESlintToRecommendedRuntimeUtils

RishhiB committed:
new lint errors

RishhiB committed:
update to strict

RishhiB committed:
Update utf-8 encoding comments to explain breaking change implications

- Clarify that utf-8 encoding identifier cannot be changed to utf8
- Explain that encoding values are stored in summaries
- Note that migration strategy needed before changing encoding format

RishhiB committed:
Update utf-8 encoding comment in driver-definitions to clarify migration path

- Add clarification that breaking change needs to be addressed first
- Make it clear that changing utf-8 to utf8 is possible but requires proper migration

jason-ha commented on packages/runtime/runtime-utils/src/compatibilityBase.ts:
```diff
@@ -68,9 +68,7 @@ export type SemanticVersion =
-		[version: MinimumMinorSemanticVersion]: T[K];
-	};
+	[K in keyof T]-?: Record<MinimumMinorSemanticVersion, T[K]>;
 };
```
Why is this kind of changed needed? What is the meaningful difference in the types?

jason-ha commented on packages/runtime/runtime-utils/src/handles.ts:
```diff
@@ -36,13 +36,15 @@ export interface ISerializedHandle {
-export const isSerializedHandle = (value: any): value is ISerializedHandle =>
-	value?.type === "__fluid_handle__";
+export const isSerializedHandle = (value: unknown): value is ISerializedHandle =>
+	// Type assertion is safe as we're only checking for the existence of the type property
```
This comment is incorrect. The check is not just for the existence of the property. It also is checking the value if it exists.

jason-ha commented on packages/runtime/runtime-utils/src/handles.ts:
```diff
@@ -120,6 +122,7 @@ export function isFluidHandle(value: unknown): value is IFluidHandle {
 	// If enableBackwardsCompatibility, run check for FluidHandles predating use of fluidHandleSymbol.
 	if (enableBackwardsCompatibility && IFluidHandle in value) {
 		// Since this check can have false positives, make it a bit more robust by checking value[IFluidHandle][IFluidHandle]
+		// Type assertion is needed for backward compatibility with old FluidHandle format
```
I don't think this comment quite makes sense.
It seems to say that if there wasn't a type assertion, then things wouldn't be backward compatible.  Does that mean that sometimes this logic is compiled against older packages that have different types?
My guess is that this logic means to convey and deal with possibility that value.IFluidHandle was in the past of type IFluidHandle. Ideally such past definitions are kept around and leveraged here instead.
